File: Ex0306_LinkedQueue/LinkedQueue.c
#include <stdio.h>
#include <stdlib.h>

typedef int datatype;

typedef struct node
{
	datatype data;
	struct node *next;
}*PtrToNode;

typedef struct listqueue
{
	PtrToNode tail;
}*LQueue;

LQueue CreateQueue(void)
{
	PtrToNode head = malloc(sizeof(struct node));
	if (head == NULL)
	{
		printf("Malloc Error.\n");
		exit(1);
	}
	head->next = head;
	LQueue Q = malloc(sizeof(struct listqueue));
	if (Q == NULL)
	{
		printf("Malloc Error.\n");
		exit(1);
	}
	Q->tail = head;
	return Q;
}


int IsEmpty(LQueue Q)
{
	if (Q == NULL || Q->tail == NULL || Q->tail->next == Q->tail)
	{
		return 1;
	}
	else
	{
		return 0;
	}
}

void Enqueue(LQueue Q, datatype x)
{
	if (Q == NULL || Q->tail == NULL)
	{
		printf("Cannot enqueue to an un-initialized queue");
		exit(1);
	}
	
	PtrToNode temp = malloc(sizeof(struct node));

	if (temp == NULL)
	{
		printf("Malloc Error.\n");
		exit(1);
	}

	temp->data = x;
	temp->next = Q->tail->next;
	Q->tail->next = temp;
	Q->tail = temp;
}

datatype Dequeue(LQueue Q)
{
	if (IsEmpty(Q))
	{
		printf("No node to dequeue.\n");
		exit(1);
	}
	
	// The node going to dequeue is the next node of head
	PtrToNode temp = Q->tail->next->next;
	datatype x = temp->data;
	
	// Check if the dequeuing node is the last node
	if (temp != Q->tail)
	{
		Q->tail->next->next = temp->next;
	}
	else
	{
		Q->tail = temp->next;
		Q->tail->next = Q->tail;
	}

	free(temp);
	temp = NULL;

	return x;
}

void DisposeQueue(LQueue Q)
{
	while (!IsEmpty(Q))
	{
		Dequeue(Q);
	}
	free(Q->tail);
	free(Q);
}

int main(void)
{
	int input[] = { 3,2,5,4,1,3,6 };
	int n = sizeof(input) / sizeof(input[0]);

	LQueue Q = CreateQueue();

	for (int i = 0; i < n; i++)
	{
		Enqueue(Q, input[i]);
	}
	
	for (int i = 0; i < n; i++)
	{
		printf("Dequeuing: %d\n", Dequeue(Q));
	}

	DisposeQueue(Q);
	Q = NULL;

	return 0;
}

File: Ex0306_LinkedQueue/LinkedQueue.java

public class LinkedQueue<E> {
    node<E> tail;

    public static class node<E> {
        private E data = null;
        private node<E> next = null;

        public node() {}

        public node(E data) {
            setData(data);
        }

        public node(E data, node<E> next) {
            this(data);
            setNext(next);
        }

        public E getData() {
            return data;
        }

        public void setData(E data) {
            this.data = data;
        }

        public node<E> getNext() {
            return next;
        }

        public boolean isLast() {
            return getNext() == null;
        }

        public void setNext(node<E> next) {
            this.next = next;
        }
    }

    public LinkedQueue() {
        makeEmpty();
    }

    public LinkedQueue(E[] inputArray) {
        this();

        for (E e: inputArray) {
            enqueue(e);
        }
    }

    public boolean isEmpty() {
        return tail.getNext() == tail;
    }

    public void makeEmpty() {
        tail = new node<>();
        tail.setNext(tail);
    }

    public void enqueue(E x) {
        node<E> newNode = new node<>(x, tail.getNext());
        tail.setNext(newNode);
        tail = newNode;
    }

    public E dequeue() {
        if (isEmpty()) {
            throw new IndexOutOfBoundsException("No node to dequeue");
        }

        // The node going to dequeue is the next node of head
        node<E> dequeuingNode = tail.getNext().getNext();
        E x = dequeuingNode.getData();

        // Check if the dequeuing node is the last node
        if (dequeuingNode != tail) {
            tail.getNext().setNext(dequeuingNode.getNext());
        }
        else {
            tail = tail.getNext();
            tail.setNext(tail);
        }

        return x;
    }
}

class TestLinkedList {
    public static void main(String[] args) {
        LinkedQueue<Integer> myQueue = new LinkedQueue<>(
                new Integer[]{3,2,5,4,1,3,6}
        );

        while(!myQueue.isEmpty()) {
            System.out.println("Dequeuing: " + myQueue.dequeue());
        }
    }
}
